File: sdk/go/sdk.go
package sdk

import (
	"crypto/x509"
	"time"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/hyperledger/fabric-gateway/pkg/identity"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"

	"github.com/nexus-lab/iot-service-blockchain/common"
)

// Sdk the iot service blockchain sdk
type Sdk struct {
	gw              *client.Gateway
	organizationId  string
	deviceId        string
	deviceRegistry  DeviceRegistryInterface
	serviceRegistry ServiceRegistryInterface
	serviceBroker   ServiceBrokerInterface
}

// SdkOptions SDK initialization options
type SdkOptions struct {
	// OrganizationId organization/MSP ID
	OrganizationId string

	// Certificate PEM-formated X509 client certificate
	Certificate []byte

	// PrivateKey PEM-formated client private key
	PrivateKey []byte

	// GatewayPeerEndpoint network address of the gateway peer
	GatewayPeerEndpoint string

	// GatewayPeerServerName server name of the gateway peer
	GatewayPeerServerName string

	// GatewayPeerTLSCertificate PEM-formated X509 TLS certificate of the gateway peer
	GatewayPeerTLSCertificate []byte

	// NetworkName blockchain network channel name
	NetworkName string

	// ChaincodeId name of the chaincode
	ChaincodeId string
}

func newIdentity(organizationId string, certificate []byte) (*identity.X509Identity, error) {
	cert, err := identity.CertificateFromPEM(certificate)
	if err != nil {
		return nil, err
	}

	id, err := identity.NewX509Identity(organizationId, cert)
	if err != nil {
		return nil, err
	}

	return id, err
}

func newSign(privateKey []byte) (identity.Sign, error) {
	key, err := identity.PrivateKeyFromPEM(privateKey)
	if err != nil {
		return nil, err
	}

	sign, err := identity.NewPrivateKeySign(key)
	if err != nil {
		return nil, err
	}

	return sign, nil
}

func newGrpcConnection(endpoint, serverName string, certificate []byte) (*grpc.ClientConn, error) {
	cert, err := identity.CertificateFromPEM(certificate)
	if err != nil {
		return nil, err
	}

	pool := x509.NewCertPool()
	pool.AddCert(cert)
	credentials := credentials.NewClientTLSFromCert(pool, serverName)

	connection, err := grpc.Dial(endpoint, grpc.WithTransportCredentials(credentials))
	if err != nil {
		return nil, err
	}

	return connection, nil
}

// NewSdk create a new SDK instance from options
func NewSdk(options *SdkOptions) (*Sdk, error) {
	sdk := new(Sdk)

	id, err := newIdentity(options.OrganizationId, options.Certificate)
	if err != nil {
		return nil, err
	}

	sign, err := newSign(options.PrivateKey)
	if err != nil {
		return nil, err
	}

	conn, err := newGrpcConnection(
		options.GatewayPeerEndpoint,
		options.GatewayPeerServerName,
		options.GatewayPeerTLSCertificate,
	)
	if err != nil {
		return nil, err
	}

	gw, err := client.Connect(
		id,
		client.WithSign(sign),
		client.WithClientConnection(conn),
		client.WithEvaluateTimeout(5*time.Second),
		client.WithEndorseTimeout(15*time.Second),
		client.WithSubmitTimeout(5*time.Second),
		client.WithCommitStatusTimeout(1*time.Minute),
	)
	if err != nil {
		return nil, err
	}

	sdk.gw = gw

	network := gw.GetNetwork(options.NetworkName)

	sdk.connectSmartContracts(network, options.ChaincodeId)
	if err = sdk.setIdentity(options.OrganizationId, options.Certificate); err != nil {
		return nil, err
	}

	return sdk, nil
}

func (s *Sdk) setIdentity(organizationId string, certificate []byte) error {
	cert, err := common.ParseCertificate(certificate)
	if err != nil {
		return err
	}

	clientId, err := common.GetClientId(cert)
	if err != nil {
		return err
	}

	s.deviceId = clientId
	s.organizationId = organizationId

	return nil
}

func (s *Sdk) connectSmartContracts(network *client.Network, chaincodeId string) {
	s.deviceRegistry = createDeviceRegistry(network, chaincodeId)
	s.serviceRegistry = createServiceRegistry(network, chaincodeId)
	s.serviceBroker = createServiceBroker(network, chaincodeId)
}

// GetDeviceId return the device/client ID of the current calling application
func (s *Sdk) GetDeviceId() string {
	return s.deviceId
}

// GetOrganizationId return the organization ID of the current calling application
func (s *Sdk) GetOrganizationId() string {
	return s.organizationId
}

// GetDeviceRegistry return the device registry
func (s *Sdk) GetDeviceRegistry() DeviceRegistryInterface {
	return s.deviceRegistry
}

// GetServiceRegistry return the service registry
func (s *Sdk) GetServiceRegistry() ServiceRegistryInterface {
	return s.serviceRegistry
}

// GetServiceBroker return the service broker
func (s *Sdk) GetServiceBroker() ServiceBrokerInterface {
	return s.serviceBroker
}

// Close close connection to the Hyperledger Fabric gateway
func (s *Sdk) Close() {
	if s.gw != nil {
		s.gw.Close()
	}
}

File: sdk/javascript/Sdk.ts
import * as grpc from '@grpc/grpc-js';
import { Gateway, Identity, Network, connect, signers } from '@hyperledger/fabric-gateway';
import * as crypto from 'crypto';
import { TextEncoder } from 'util';

import { DeviceRegistryInterface, createDeviceRegistry } from './DeviceRegistry';
import { ServiceBrokerInterface, createServiceBroker } from './ServiceBroker';
import { ServiceRegistryInterface, createServiceRegistry } from './ServiceRegistry';
import { getClientId, parseCertificate } from './identity';

/**
 * SDK initialization options
 */
export interface SdkOptions {
  /**
   * Organization/MSP ID
   */
  organizationId: string;

  /**
   * PEM-formated X509 client certificate
   */
  certificate: string;

  /**
   * PEM-formated client private key
   */
  privateKey: string;

  /**
   * Network address of the gateway peer
   */
  gatewayPeerEndpoint: string;

  /**
   * Server name of the gateway peer
   */
  gatewayPeerServerName: string;

  /**
   * PEM-formated X509 TLS certificate of the gateway peer
   */
  gatewayPeerTLSCertificate: string;

  /**
   * Blockchain network channel name
   */
  networkName: string;

  /**
   * Name of the chaincode
   */
  chaincodeId: string;
}

function newIdentity(organizationId: string, certificate: string): Identity {
  const encoder = new TextEncoder();
  return {
    mspId: organizationId,
    credentials: encoder.encode(certificate),
  };
}

function newSigner(privateKey: string) {
  return signers.newPrivateKeySigner(crypto.createPrivateKey(Buffer.from(privateKey)));
}

function newGrpcConnection(endpoint: string, serverName: string, certificate: string) {
  const credentials = grpc.credentials.createSsl(Buffer.from(certificate));
  return new grpc.Client(endpoint, credentials, { 'grpc.ssl_target_name_override': serverName });
}

/**
 * The IoT service blockchain sdk
 */
export default class Sdk {
  private grpcConnection: grpc.Client;
  private gateway: Gateway;
  private organizationId!: string;
  private deviceId!: string;
  private deviceRegistry!: DeviceRegistryInterface;
  private serviceRegistry!: ServiceRegistryInterface;
  private serviceBroker!: ServiceBrokerInterface;

  /**
   * @param options SDK initialization options
   */
  constructor(options: SdkOptions) {
    this.grpcConnection = newGrpcConnection(
      options.gatewayPeerEndpoint,
      options.gatewayPeerServerName,
      options.gatewayPeerTLSCertificate,
    );
    const identity = newIdentity(options.organizationId, options.certificate);
    const signer = newSigner(options.privateKey);

    this.gateway = connect({
      client: this.grpcConnection,
      identity,
      signer,
      evaluateOptions: () => ({ deadline: Date.now() + 5000 }), // 5 seconds
      endorseOptions: () => ({ deadline: Date.now() + 15000 }), // 15 seconds
      submitOptions: () => ({ deadline: Date.now() + 5000 }), // 5 seconds
      commitStatusOptions: () => ({ deadline: Date.now() + 60000 }), // 1 minute
    });

    const network = this.gateway.getNetwork(options.networkName);
    this.connectSmartContracts(network, options.chaincodeId);
    this.setIdentity(options.organizationId, options.certificate);
  }

  private setIdentity(organizationId: string, certificate: string) {
    this.deviceId = getClientId(parseCertificate(certificate));
    this.organizationId = organizationId;
  }

  private connectSmartContracts(network: Network, chaincodeId: string) {
    this.deviceRegistry = createDeviceRegistry(network, chaincodeId);
    this.serviceRegistry = createServiceRegistry(network, chaincodeId);
    this.serviceBroker = createServiceBroker(network, chaincodeId);
  }

  /**
   * Get the device/client ID of the current calling application
   *
   * @returns the device/client ID of the current calling application
   */
  getDeviceId() {
    return this.deviceId;
  }

  /**
   * Get the organization ID of the current calling application
   *
   * @returns the organization ID of the current calling application
   */
  getOrganizationId() {
    return this.organizationId;
  }

  /**
   * Get the device registry
   *
   * @returns the device registry
   */
  getDeviceRegistry() {
    return this.deviceRegistry;
  }

  /**
   * Get the server registry
   *
   * @returns the server registry
   */
  getServiceRegistry() {
    return this.serviceRegistry;
  }

  /**
   * Get the service broker
   *
   * @returns the service broker
   */
  getServiceBroker() {
    return this.serviceBroker;
  }

  /**
   * Close connection to the Hyperledger Fabric gateway
   */
  close() {
    this.gateway.close();
    this.grpcConnection.close();
  }
}
